support/vm/vector_gen.go

```diff
@@ -0,0 +1,407 @@
+	opts = append(opts, SetID(id))
+
+	return opts
+}
+
+//
+// Internal types for serialization
+// Taken from https://github.com/filecoin-project/test-vectors/blob/master/schema/schema.go
+//
+
+type generationData struct {
+	Source string `json:"source"`
+}
+
+type metadata struct {
+	ID  string           `json:"id"`
+	Gen []generationData `json:"gen"`
+}
+
+type variant struct {
+	// ID of the variant, usually the codename of the upgrade.
+	ID             string `json:"id"`
+	Epoch          int64  `json:"epoch"`
+	NetworkVersion uint   `json:"nv"`
+}
+
+type preconditions struct {
+	Variants   []variant        `json:"variants"`
+	StateTree  *stateTreeSerial `json:"state_tree,omitempty"`
+	BaseFee    *gbig.Int        `json:"basefee,omitempty"`
+	CircSupply *gbig.Int        `json:"circ_supply,omitempty"`
+}
+
+type base64EncodedBytes []byte
+
+func (b base64EncodedBytes) String() string {
+	return base64.StdEncoding.EncodeToString(b)
+}
+
+// MarshalJSON implements json.Marshal for Base64EncodedBytes
+func (b base64EncodedBytes) MarshalJSON() ([]byte, error) {
+	return json.Marshal(b.String())
+}
+
+type messageSerial struct {
+	Bytes base64EncodedBytes `json:"bytes"`
+}
+type stateTreeSerial struct {
+	RootCID cid.Cid `json:"root_cid"`
+}
+
+// Receipt represents a receipt to match against.
+type receiptSerial struct {
+	// ExitCode must be interpreted by the driver as an exitcode.ExitCode
+	// in Lotus, or equivalent type in other implementations.
+	ExitCode    int64              `json:"exit_code"`
+	ReturnValue base64EncodedBytes `json:"return"`
+	GasUsed     int64              `json:"gas_used"`
+}
+
+// Postconditions contain a representation of VM state at th end of the test
+type postconditions struct {
+	StateTree *stateTreeSerial `json:"state_tree"`
+	Receipts  []*receiptSerial `json:"receipts"`
+}
+
+type testVectorSerial struct {
+	Class string `json:"class"`
+
+	Meta *metadata `json:"_meta"`
+
+	// CAR binary data to be loaded into the test environment. Should
+	// contain objects of entire state tree
+	CAR base64EncodedBytes `json:"car"`
+
+	Pre *preconditions `json:"preconditions"`
+
+	ApplyMessages []messageSerial `json:"apply_messages,omitempty"`
+
+	Post *postconditions `json:"postconditions"`
+}
+
+func newTestVectorSerial(tv *TestVector) (*testVectorSerial, error) {
+	zero := big.Zero()
+	circSupply := tv.CircSupply
+	var msgBuf bytes.Buffer
+	if err := tv.Message.MarshalCBOR(&msgBuf); err != nil {
+		return nil, err
+	}
+	msgBytes := msgBuf.Bytes()
+	var retBuf bytes.Buffer
+	if err := tv.Receipt.Ret.MarshalCBOR(&retBuf); err != nil {
+		return nil, err
+	}
+	retBytes := retBuf.Bytes()
+
+	return &testVectorSerial{
+		Class: "message",
+		Meta: &metadata{
+			ID: tv.ID,
+			Gen: []generationData{
+				{Source: "specs-actors_test_auto_gen"},
+			},
+		},
+		CAR: tv.StartState,
+		Pre: &preconditions{
+			Variants: []variant{
+				{ID: defaultNetworkName, Epoch: int64(tv.Epoch), NetworkVersion: uint(tv.Version)},
```

Comment:
Excellent callout.  I did not exactly inherit this.  The main reasons for using this serialization type were to separate the complexity of what the actors generation vectors need from the full scope of the test vector format.  For example it is really nice to not have to think about having multiple test-vector variants, or any non-message class vectors at the actors generation level.

We can build good general purpose patterns for working with the full format to avoid relying on assumptions to make static shortcuts but redoing these here without taking on the full responsibility of the test-vector format would have been overkill.

My plan for the moment is to actually couple these types further into the specs-actors reference vm to prevent any external reliance on these internal shortcut assumptions.  Then proper decoupling can be done in one go along with introducing a generally/externally usable format: https://github.com/filecoin-project/specs-actors/issues/1455